modules/packages/Sort.chpl

```diff
@@ -18,53 +18,321 @@
+  proc Comparator.key(a) { return abs(a); }
+
+  var absComparator: Comparator;
+
+  QuickSort(Array, comparator=absComparator);
+
+  // This will output: -1, 2, 3, -4
+  writeln(Array);
+
+
+Compare Comparator
+~~~~~~~~~~~~~~~~~~
+
+The ``compare(a, b)`` method accepts 2 arguments, which will be 2 elements from
+the array being sorted. The return value should be an integer indicating how a
+and b compare to each other, as shown below:
+
+  * > 0 if ``a > b``
+  * 0 if ``a == b``
+  * < 0 if ``a < b``
+
+The default compare method for a numeric signed type would look like this:
+
+.. code-block:: chapel
+
+    proc DefaultComparator.compare(a, b) {
+      return a - b;
+    }
+
+
+The absolute value comparison example from above can alternatively be
+implemented with a compare method:
+
+.. code-block:: chapel
+
+  var Array = [-1, -4, 2, 3];
+
+  // Empty record serves as comparator
+  record Comparator { }
+
+  // compare method defines how 2 elements are compared
+  proc Comparator.compare(a, b) {
+    return abs(a) - abs(b);
+  }
+
+  var absComparator: Comparator;
+
+  QuickSort(Array, comparator=absComparator);
+
+  // This will output: -1, 2, 3, -4
+  writeln(Array);
+
+Reverse Comparator
+~~~~~~~~~~~~~~~~~~
+
+Sort functions in Chapel do not have a ``reverse`` argument. Instead, reverse
+sorting is handled through the comparator interface.
+
+A module-defined :const:`reversecomparator` can be passed to a sort function to
+reverse the default sorting order.
+
+.. code-block:: chapel
+
+  var Array = [-1, -4, 2, 3];
+
+  // Using module-defined 'reversecomparator'
+  QuickSort(Array, comprator=reversecomparator)
+
+  // This will output: 3, 2, -1, -4
+  writeln(Array);
+
+
+To reverse the sort order of a user-defined comparator, pass the user-defined
+comparator to the constructor of the module-defined
+:record:`ReverseComparator` record, which can be passed to the sort function.
+
+.. code-block:: chapel
+
+  var Array = [-1, -4, 2, 3];
+
+  // Empty record serves as comparator
+  record Comparator { }
+
+  // compare method defines how 2 elements are compared
+  proc Comparator.compare(a, b) {
+    return abs(a) - abs(b);
+  }
+
+  var absReverseComparator: ReverseComparator(Comparator);
+
+  QuickSort(Array, comparator=absReverseComparator);
+
+  // This will output: -4, 3, 2, -1
+  writeln(Array);
 
  */
 module Sort {
 
-private inline proc chpl_sort_cmp(a, b, param reverse=false, param eq=false) {
-  if eq {
-    if reverse then return a >= b;
-    else return a <= b;
+
+/* Module-defined comparators */
+
+/*
+  Instance of :record:`DefaultComparator` used as default ``comparator=``
+  argument when no comparator is passed to a sort function
+*/
+const defaultcomparator: DefaultComparator;
+
+
+
+/*
+   Instance of :record:`ReverseComparator`. Pass this as the ``comparator=``
+   argument of a sort function to reverse the sort order.
+ */
+const reversecomparator: ReverseComparator(DefaultComparator);
```

Comment:
Why not 'reverseComparator' ?
'reverseOrder' is another possible name for this, as is 'reverseDefaultComparator'.